File: src/baseline/sw_baseline.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define MAX(A,B) ((A) > (B) ? (A) : (B))
#define MATCH     2
#define MISMATCH -1
#define GAP      -2

void generate_sequence(char *seq, int n) {
    const char alphabet[] = "ACGT";
    for (int i = 0; i < n; i++)
        seq[i] = alphabet[rand() % 4];
    seq[n] = '\0';
}

int smith_waterman(const char *seq1, const char *seq2, int len1, int len2) {
    int **H = malloc((len1 + 1) * sizeof(int *));
    for (int i = 0; i <= len1; i++)
        H[i] = calloc(len2 + 1, sizeof(int));

    int max_score = 0;
    for (int i = 1; i <= len1; i++) {
        for (int j = 1; j <= len2; j++) {
            int match = H[i - 1][j - 1] + (seq1[i - 1] == seq2[j - 1] ? MATCH : MISMATCH);
            int del = H[i - 1][j] + GAP;
            int ins = H[i][j - 1] + GAP;
            H[i][j] = MAX(0, MAX(match, MAX(del, ins)));
            if (H[i][j] > max_score)
                max_score = H[i][j];
        }
    }

    for (int i = 0; i <= len1; i++) free(H[i]);
    free(H);

    return max_score;
}

int main(int argc, char **argv) {
    if (argc < 3) {
        fprintf(stderr, "Usage: %s <sequence_length> <num_of_threads>\n", argv[0]);
        return 1;
    }

    int N = atoi(argv[1]);
    srand(42);

    char *seq1 = malloc((N + 1) * sizeof(char));
    char *seq2 = malloc((N + 1) * sizeof(char));

    generate_sequence(seq1, N);
    generate_sequence(seq2, N);

    clock_t start = clock();
    int score = smith_waterman(seq1, seq2, N, N);
    clock_t end = clock();

    double elapsed = (double)(end - start) / CLOCKS_PER_SEC;
    printf("Sequence length: %d\n", N);
    printf("Smith-Waterman score: %d\n", score);
    printf("Execution time: %.6f seconds\n", elapsed);

    free(seq1);
    free(seq2);
    return 0;
}

File: src/baseline/sw_baseline.py
import sys
import random
import time

MATCH = 2
MISMATCH = -1
GAP = -2

def generate_sequence(n):
    return ''.join(random.choice('ACGT') for _ in range(n))

def smith_waterman(seq1, seq2):
    len1, len2 = len(seq1), len(seq2)
    H = [[0]*(len2+1) for _ in range(len1+1)]
    max_score = 0

    for i in range(1, len1+1):
        for j in range(1, len2+1):
            match = H[i-1][j-1] + (MATCH if seq1[i-1] == seq2[j-1] else MISMATCH)
            delete = H[i-1][j] + GAP
            insert = H[i][j-1] + GAP
            H[i][j] = max(0, match, delete, insert)
            if H[i][j] > max_score:
                max_score = H[i][j]
    return max_score

if __name__ == "__main__":
    if len(sys.argv) < 3:
        print("Usage: python3 sw_baseline.py <sequence_length> <num_of_threads>")
        sys.exit(1)

    N = int(sys.argv[1])
    random.seed(42)
    seq1 = generate_sequence(N)
    seq2 = generate_sequence(N)

    start = time.time()
    score = smith_waterman(seq1, seq2)
    end = time.time()

    print(f"Sequence length: {N}")
    print(f"Smith-Waterman score: {score}")
    print(f"Execution time: {end - start:.6f} seconds")
